packages/core-transaction-pool/src/query.ts

```diff
@@ -0,0 +1,156 @@
+import { Container, Contracts } from "@arkecosystem/core-kernel";
+import { Interfaces, Transactions } from "@arkecosystem/crypto";
+
+import { IteratorMany } from "./utils";
+
+class QueryIterator implements Iterator<Interfaces.ITransaction> {
+    private readonly iter: Iterator<Interfaces.ITransaction>;
+    private readonly predicate: Contracts.TransactionPool.Predicate | undefined;
+
+    public constructor(iter: Iterator<Interfaces.ITransaction>, predicate?: Contracts.TransactionPool.Predicate) {
+        this.iter = iter;
+        this.predicate = predicate;
+    }
+
+    public next(value?: any): IteratorResult<Interfaces.ITransaction> {
+        for (let r = this.iter.next(); !r.done; r = this.iter.next()) {
+            if (!this.predicate || this.predicate(r.value)) {
+                return r;
+            }
+        }
+
+        return { done: true, value: undefined };
```

Comment:
We should avoid indicators like `done/success/failure` based on boolean values. It should be possible to determine the success or failure of a method solely based on its return value without any additional indicators.

Reply:
This is ES iterator interface, not mine. This is what `for (const value of iterable)` expects.

Reply:
It is gone now